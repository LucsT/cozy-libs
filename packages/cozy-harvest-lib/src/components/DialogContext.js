import { createContext, useContext } from 'react'

const DialogContext = createContext({})
export const useDialogContext = () => {
  return useContext(DialogContext)
}
export default DialogContext
