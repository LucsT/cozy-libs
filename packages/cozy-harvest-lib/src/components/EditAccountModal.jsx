import React, { Component } from 'react'
import PropTypes from 'prop-types'

import get from 'lodash/get'
import flow from 'lodash/flow'
import { withClient } from 'cozy-client'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Modal, {
  ModalContent,
  ModalHeader
} from 'cozy-ui/transpiled/react/Modal'
import CipherIcon from 'cozy-ui/transpiled/react/CipherIcon'

import { fetchAccount } from '../connections/accounts'
import * as triggersModel from '../helpers/triggers'
import TriggerManager from './TriggerManager'
import { withMountPointPushHistory } from './MountPointContext'
import logger from '../logger'

export class EditAccountModal extends Component {
  constructor(props) {
    super(props)
    this.state = {
      trigger: null,
      account: null,
      fetching: true,
      error: false
    }

    this.redirectToAccount = this.redirectToAccount.bind(this)
  }

  componentDidMount() {
    const { accountId, accounts } = this.props

    /**
     * @TODO In theory we can have several trigger for the same account.
     * If so this code will not work as excepted. This case is theoretical
     */
    const matchingTrigger = get(
      accounts.find(account => account.account._id === accountId),
      'trigger'
    )
    if (matchingTrigger) {
      this.fetchAccount(matchingTrigger)
    } else {
      logger.warn('No matching trigger for account', accountId)
    }
  }
  /**
   * TODO use queryConnect to know if we're fecthing or not
   */
  async fetchAccount(trigger) {
    const { client } = this.props
    this.setState({ fetching: true })

    try {
      const accountId = triggersModel.getAccountId(trigger)
      const account = await fetchAccount(client, accountId)
      this.setState({
        account,
        trigger
      })
      return account
    } catch (error) {
      this.setState({
        error
      })
    } finally {
      this.setState({
        fetching: false
      })
    }
  }

  redirectToAccount() {
    const { account } = this.state
    if (account) {
      this.props.pushHistory(`/accounts/${account._id}`)
    } else {
      this.props.pushHistory(`/accounts`)
    }
  }

  render() {
    /**
     * We don't use the dismiss action pros that we can have from our
     * Routes component since this modal has to be on top on the previous one
     * So when we quit it, we have to go back to the previous one.
     *
     * When we are on mobile, we display a back button
     * On desktop we display a cross
     */
    const { konnector } = this.props
    const { trigger, account, fetching } = this.state
    return (
      <Modal
        dismissAction={this.redirectToAccount}
        mobileFullscreen
        size="small"
      >
        <ModalHeader
          title={
            <div className="u-flex u-flex-items-center">
              <CipherIcon konnector={konnector.slug} className="u-mr-1" />
              {konnector.name}
            </div>
          }
        />
        <ModalContent>
          {fetching ? (
            <div className="u-pv-2 u-ta-center">
              <Spinner size="xxlarge" />
            </div>
          ) : (
            <TriggerManager
              account={account}
              konnector={konnector}
              initialTrigger={trigger}
              onSuccess={this.redirectToAccount}
              showError={true}
              onVaultDismiss={this.redirectToAccount}
            />
          )}
        </ModalContent>
      </Modal>
    )
  }
}

EditAccountModal.propTypes = {
  konnector: PropTypes.object.isRequired,
  accountId: PropTypes.string.isRequired,
  accounts: PropTypes.array.isRequired
}

export default flow(
  withClient,
  withMountPointPushHistory
)(EditAccountModal)
