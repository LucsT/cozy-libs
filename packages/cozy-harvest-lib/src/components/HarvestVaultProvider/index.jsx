import React from 'react'
import PropTypes from 'prop-types'
import { withClient } from 'cozy-client'
import { VaultProvider } from 'cozy-keys-lib'

class HarvestVaultProvider extends React.PureComponent {
  render() {
    const { children, client } = this.props

    const vaultInstanceUrl = client.getStackClient().uri

    return <VaultProvider instance={vaultInstanceUrl}>{children}</VaultProvider>
  }
}

HarvestVaultProvider.propTypes = {
  client: PropTypes.object
}

export default withClient(HarvestVaultProvider)
