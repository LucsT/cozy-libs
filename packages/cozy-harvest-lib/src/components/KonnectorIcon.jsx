import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { withClient } from 'cozy-client'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'

// TODO move this to cozy-ui
class KonnectorIcon extends PureComponent {
  fetchIcon() {
    const { konnector, client } = this.props
    return client.getStackClient().getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  render() {
    // eslint-disable-next-line no-unused-vars
    const { client, konnector, ...restProps } = this.props
    return <AppIcon fetchIcon={this.fetchIcon.bind(this)} {...restProps} />
  }
}

KonnectorIcon.propTypes = {
  konnector: PropTypes.object.isRequired,
  client: PropTypes.object.isRequired
}
export default withClient(KonnectorIcon)
