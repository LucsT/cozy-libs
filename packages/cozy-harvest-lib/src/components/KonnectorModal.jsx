import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withMutations } from 'cozy-client'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Button from 'cozy-ui/transpiled/react/Button'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Modal, {
  ModalContent,
  ModalHeader
} from 'cozy-ui/transpiled/react/Modal'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Icon from 'cozy-ui/transpiled/react/Icon'
import { Text } from 'cozy-ui/transpiled/react/Text'
import get from 'lodash/get'

import accountMutations from '../connections/accounts'
import triggersMutations from '../connections/triggers'
import * as triggersModel from '../helpers/triggers'

import withLocales from './hoc/withLocales'

import AccountSelectBox from './AccountSelectBox/AccountSelectBox'
import AccountList from './AccountList/AccountList'
import KonnectorConfiguration from './KonnectorConfiguration/KonnectorConfiguration'

/**
 * KonnectorModal open a Modal related to a given konnector. It fetches the
 * first account and then include a TriggerManager component.
 *
 * This component is aimed to offer an UI to manage all the konnector related
 * triggers and accounts.
 */
export class KonnectorModal extends PureComponent {
  state = {
    account: null,
    fetching: false,
    trigger: null,
    accounts: []
  }

  constructor(props) {
    super(props)
    this.fetchIcon = this.fetchIcon.bind(this)
  }
  /**
   * TODO We should not fetchAccounts and fetchAccount since we already have the informations
   * in the props. We kept this sytem for compatibility on the existing override.
   * Next tasks: remove these methods and rewrite the override
   */
  async componentDidMount() {
    await this.fetchAccounts()
    const { accounts } = this.state

    if (this.props.accountId) this.loadSelectedAccountId()
    else if (accounts.length === 1)
      this.requestAccountChange(accounts[0].account, accounts[0].trigger)
  }

  componentWillUnmount() {
    const { into } = this.props
    if (!into) return
    // The Modal is never closed after a dismiss on Preact apps, even if it is
    // not rendered anymore. The best hack we found is to explicitly empty the
    // modal portal container.
    setTimeout(() => {
      try {
        const modalRoot = document.querySelector(into)
        modalRoot.innerHTML = ''
        // eslint-disable-next-line no-empty
      } catch (error) {}
    }, 50)
  }

  componentDidUpdate(prevProps) {
    if (this.props.accountId && this.props.accountId !== prevProps.accountId) {
      this.loadSelectedAccountId()
    }
  }

  requestAccountChange(account, trigger) {
    // This component can either defer the account switching to a parent component through the onAccountChange prop, or handle the change itself if the prop is missing
    const { onAccountChange } = this.props
    onAccountChange ? onAccountChange(account) : this.fetchAccount(trigger)
  }

  loadSelectedAccountId() {
    const selectedAccountId = this.props.accountId
    const { accounts } = this.state

    const matchingTrigger = get(
      accounts.find(account => account.account._id === selectedAccountId),
      'trigger'
    )

    if (matchingTrigger) this.fetchAccount(matchingTrigger)
  }

  async fetchAccounts() {
    const triggers = this.props.konnector.triggers.data
    const { findAccount } = this.props
    const accounts = await Promise.all(
      triggers.map(async trigger => {
        return {
          account: await findAccount(triggersModel.getAccountId(trigger)),
          trigger
        }
      })
    )
    this.setState({ accounts })
  }

  async fetchAccount(trigger) {
    const { findAccount } = this.props
    this.setState({ fetching: true })

    try {
      const account = await findAccount(triggersModel.getAccountId(trigger))
      this.setState({
        account,
        trigger
      })
    } catch (error) {
      this.setState({
        error
      })
    } finally {
      this.setState({
        fetching: false
      })
    }
  }

  fetchIcon() {
    const { client } = this.context
    const { konnector } = this.props
    return client.stackClient.getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  async refetchTrigger() {
    const { fetchTrigger } = this.props
    const { trigger } = this.state

    const upToDateTrigger = await fetchTrigger(trigger._id)
    this.setState({
      trigger: upToDateTrigger
    })
  }

  render() {
    const { dismissAction, konnector, into, t, createAction } = this.props
    const { account, accounts } = this.state

    return (
      <Modal
        dismissAction={dismissAction}
        mobileFullscreen
        size="small"
        into={into}
        closable={false}
      >
        <ModalHeader className="u-pr-2">
          <div className="u-flex u-flex-row u-w-100 u-flex-items-center">
            <div className="u-w-3 u-h-3 u-mr-half">
              <AppIcon fetchIcon={this.fetchIcon} />
            </div>
            <div className="u-flex-grow-1 u-mr-half">
              <h3 className="u-title-h3 u-m-0">{konnector.name}</h3>

              {accounts.length === 0 && (
                <Text className="u-slateGrey">{t('loading')}</Text>
              )}
              {accounts.length > 0 && account && (
                <AccountSelectBox
                  selectedAccount={account}
                  accountList={accounts}
                  onChange={option => {
                    this.requestAccountChange(option.account, option.trigger)
                  }}
                  onCreate={createAction}
                />
              )}
            </div>
            <Button
              icon={<Icon icon={'cross'} size={'24'} />}
              onClick={dismissAction}
              iconOnly
              label={t('close')}
              subtle
              theme={'secondary'}
            />
          </div>
        </ModalHeader>
        <ModalContent>{this.renderModalContent()}</ModalContent>
      </Modal>
    )
  }

  renderModalContent() {
    const { dismissAction, konnector, t, createAction } = this.props
    const { account, accounts, error, fetching, trigger } = this.state

    if (fetching) {
      return (
        <Spinner
          size="xxlarge"
          className="u-flex u-flex-justify-center u-pv-3"
        />
      )
    } else if (error) {
      return (
        <Infos
          actionButton={
            <Button theme="danger">{t('modal.konnector.error.button')}</Button>
          }
          title={t('modal.konnector.error.title')}
          text={t('modal.konnector.error.description', error)}
          isImportant
        />
      )
    } else if (!account) {
      return (
        <AccountList
          accounts={accounts}
          konnector={konnector}
          onPick={option => {
            this.requestAccountChange(option.account, option.trigger)
          }}
          addAccount={createAction}
        />
      )
    } else {
      return (
        <KonnectorConfiguration
          konnector={konnector}
          trigger={trigger}
          account={account}
          onAccountDeleted={dismissAction}
          addAccount={createAction}
          refetchTrigger={this.refetchTrigger}
        />
      )
    }
  }
}

KonnectorModal.propTypes = {
  into: PropTypes.string,
  konnector: PropTypes.shape({
    slug: PropTypes.string,
    name: PropTypes.string,
    vendor_link: PropTypes.string,
    triggers: PropTypes.shape({
      data: PropTypes.arrayOf(PropTypes.object)
    })
  }),
  findAccount: PropTypes.func,
  fetchTrigger: PropTypes.func,
  dismissAction: PropTypes.func,
  createAction: PropTypes.func,
  onAccountChange: PropTypes.func,
  t: PropTypes.func
}

KonnectorModal.contextTypes = {
  client: PropTypes.object.isRequired
}

export default withMutations(accountMutations, triggersMutations)(
  withLocales(KonnectorModal)
)
