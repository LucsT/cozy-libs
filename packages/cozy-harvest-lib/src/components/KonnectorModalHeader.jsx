import React from 'react'
import PropTypes from 'prop-types'
import { ModalHeader } from 'cozy-ui/transpiled/react/Modal'
import KonnectorIcon from './KonnectorIcon'
import { Media, Img } from 'cozy-ui/transpiled/react/Media'

const KonnectorModalHeader = ({ konnector, children }) => {
  return (
    <ModalHeader className="u-pr-2">
      <Media>
        <Img className="u-w-3 u-h-3 u-mr-1">
          <KonnectorIcon konnector={konnector} />
        </Img>
        <div className="u-flex-grow-1 u-mr-1">
          <h3 className="u-title-h3 u-m-0">{konnector.name}</h3>
          {children}
        </div>
      </Media>
    </ModalHeader>
  )
}

KonnectorModalHeader.propTypes = {
  konnector: PropTypes.object.isRequired,
  children: PropTypes.node
}

export default KonnectorModalHeader
