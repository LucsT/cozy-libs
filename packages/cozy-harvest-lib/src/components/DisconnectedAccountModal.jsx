import React, { useRef, useState } from 'react'
import { Contracts } from 'cozy-harvest-lib/dist/components/KonnectorConfiguration/ConfigurationTab/Contracts'
import KonnectorModalHeader from 'cozy-harvest-lib/dist/components/KonnectorModalHeader'

import { getCreatedByApp } from 'cozy-client/dist/models/utils'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import Modal, { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import Card from 'cozy-ui/transpiled/react/Card'
import {
  Tabs,
  Tab,
  TabList,
  TabPanels,
  TabPanel
} from 'cozy-ui/transpiled/react/Tabs'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

import withLocales from './hoc/withLocales'
import { getAccountInstitutionLabel } from './KonnectorConfiguration/ConfigurationTab/bankAccountHelpers'

const createDummyKonnectorFromAccount = account => {
  return {
    name: getAccountInstitutionLabel(account),
    slug: getCreatedByApp(account) || null
  }
}

/**
 * Serves when configuring vendor accounts whose io.cozy.accounts
 * has been deleted.
 *
 * The UI must be as close as possible to the AccountModal but since
 * we lack a lot of data to show an AccountModal (no trigger, no konnector,
 * no account), it was implemented as a separate component.
 */
const DisconnectedModal = ({ accounts, onClose, initialActiveTab }) => {
  const { t } = useI18n()
  // We keep the konnector in a ref so that when we remove all accounts,
  // we still have a konnector to show the icon
  const konnectorRef = useRef()
  if (!konnectorRef.current) {
    konnectorRef.current = createDummyKonnectorFromAccount(accounts[0])
  }
  const { isMobile } = useBreakpoints()

  const [activeTab, setActiveTab] = useState(initialActiveTab)

  return (
    <Modal mobileFullscreen={true} dismissAction={onClose}>
      <KonnectorModalHeader konnector={konnectorRef.current} />
      <ModalContent className={isMobile ? 'u-p-0' : null}>
        <Tabs initialActiveTab={initialActiveTab}>
          <TabList>
            <Tab onClick={() => setActiveTab('data')} name="data">
              {t('modal.tabs.data')}
            </Tab>
            <Tab
              onClick={() => setActiveTab('configuration')}
              name="configuration"
            >
              {t('modal.tabs.configuration')}
            </Tab>
          </TabList>
        </Tabs>
      </ModalContent>
      <TabPanels activeTab={activeTab}>
        <TabPanel name="data" className="u-pt-1">
          <ModalContent>
            <Card>{t('disconnectedAccountModal.disconnected-help')}</Card>
          </ModalContent>
        </TabPanel>
        <TabPanel name="configuration">
          {accounts.length ? (
            <ModalContent className={isMobile ? 'u-p-0' : null}>
              <Contracts contracts={accounts} />
            </ModalContent>
          ) : (
            t('contracts.no-contracts')
          )}
        </TabPanel>
      </TabPanels>
    </Modal>
  )
}

DisconnectedModal.defaultProps = {
  initialActiveTab: 'configuration'
}

export default withLocales(DisconnectedModal)
