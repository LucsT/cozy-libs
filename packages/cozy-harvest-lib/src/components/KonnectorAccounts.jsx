import React from 'react'
import PropTypes from 'prop-types'
import { withClient, withMutations } from 'cozy-client'
import { withRouter } from 'react-router'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import { translate } from 'cozy-ui/transpiled/react/I18n'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Button from 'cozy-ui/transpiled/react/Button'
import get from 'lodash/get'
import compose from 'lodash/flowRight'
import keyBy from 'lodash/keyBy'
import CozyRealtime from 'cozy-realtime'

import { findAccount } from '../connections/accounts'
import triggersMutations from '../connections/triggers'
import * as triggersModel from '../helpers/triggers'
import KonnectorModalHeader from './KonnectorModalHeader'

/**
 * Returns { trigger, account } list
 */
const fetchAccountsFromTriggers = async (client, triggers) => {
  const accountCol = client.collection('io.cozy.accounts')
  const accountIdToTrigger = keyBy(triggers, triggersModel.getAccountId)
  const accountIds = Object.keys(accountIdToTrigger)
  const { data: accounts } = await accountCol.getAll(accountIds)
  return accounts
    .filter(Boolean)
    .map(account => ({ account, trigger: accountIdToTrigger[account._id] }))
}

export class KonnectorAccounts extends React.Component {
  constructor(props) {
    super(props)
    this.realtime = new CozyRealtime({ client: this.props.client })
    this.state = {
      fetchingAccounts: true,
      error: null,
      accounts: []
    }
  }

  /**
   * At the moment we fetch the accounts manually and the triggers are received as props
   * TODO Use queryConnect to fetch the accounts and the triggers once the home uses queryConnect too
   * (Using queryConnect here before the home would result in duplicate queries and longer loading times than necessary)
   */
  async componentDidMount() {
    await this.fetchAccounts()

    await this.realtime.subscribe(
      'updated',
      'io.cozy.jobs',
      this.handleTriggerUpdate.bind(this)
    )
  }

  componentDidUpdate(prevProps) {
    // After leaving the new account screen, we need to refetch accounts in case a new one was created
    if (
      this.props.location.pathname !== prevProps.location.pathname &&
      /\/new\/?$/.test(prevProps.location.pathname)
    ) {
      this.fetchAccounts()
    }
  }

  async handleTriggerUpdate(job) {
    const { fetchTrigger } = this.props
    const { accounts } = this.state
    const triggerId = job.trigger_id

    const matchingAccount = accounts.find(
      ({ trigger }) => trigger && trigger._id === triggerId
    )

    if (matchingAccount) {
      const trigger = await fetchTrigger(triggerId)
      const updatedAccountIndex = accounts.indexOf(matchingAccount)

      this.setState({
        accounts: [
          ...accounts.slice(0, updatedAccountIndex),
          {
            account: matchingAccount.account,
            trigger
          },
          ...accounts.slice(updatedAccountIndex + 1)
        ]
      })
    }
  }

  async fetchAccounts() {
    const triggers = get(this.props, 'konnector.triggers.data', [])
    const { client } = this.props
    this.setState({ fetchingAccounts: true })
    try {
      const triggerAccounts = await fetchAccountsFromTriggers(client, triggers)
      this.setState({ accounts: triggerAccounts, error: null })
    } catch (error) {
      console.error(error)
      this.setState({ error })
    } finally {
      this.setState({ fetchingAccounts: false })
    }
  }

  render() {
    const { accounts, fetchingAccounts, error } = this.state
    const { konnector, t } = this.props
    return (
      <>
        {(error || fetchingAccounts) && (
          <KonnectorModalHeader konnector={konnector} />
        )}
        {error && (
          <ModalContent>
            <Infos
              actionButton={
                <Button
                  theme="danger"
                  onClick={this.fetchAccounts.bind(this)}
                  label={t('modal.accounts.error.retry')}
                />
              }
              title={t('modal.accounts.error.title')}
              text={t('modal.accounts.error.description')}
              icon="warning"
              isImportant
            />
          </ModalContent>
        )}
        {fetchingAccounts && (
          <ModalContent>
            <div className="u-pv-2 u-ta-center">
              <Spinner size="xxlarge" />
            </div>
          </ModalContent>
        )}
        {!error && !fetchingAccounts && this.props.children(accounts)}
      </>
    )
  }
}

KonnectorAccounts.propTypes = {
  konnector: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  client: PropTypes.object.isRequired,
  fetchTrigger: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
}

export default compose(
  withMutations(triggersMutations),
  withRouter,
  translate(),
  withClient
)(KonnectorAccounts)
