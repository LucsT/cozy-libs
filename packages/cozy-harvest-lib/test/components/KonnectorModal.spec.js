/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'

import { KonnectorModal } from 'components/KonnectorModal'
import TriggerManager from 'components/TriggerManager'

const t = key => key

const findAccountMock = jest.fn().mockImplementation(() => ({
  _id: '123',
  doctype: 'io.cozy.accounts'
}))

describe('KonnectorModal', () => {
  const mockKonnector = {
    slug: 'mock',
    name: 'Mock',
    triggers: {
      data: [{ _id: 784, doctype: 'io.cozy.triggers' }]
    }
  }
  const props = {
    findAccount: findAccountMock,
    dismissAction: jest.fn(),
    onSuccess: jest.fn(),
    konnector: mockKonnector,
    t
  }
  const shallowOptions = {
    context: {
      client: {}
    }
  }

  it('should show a spinner while loading', () => {
    const component = shallow(<KonnectorModal {...props} />, shallowOptions)

    const spinner = component
      .dive()
      .find('ModalContent')
      .childAt(0)
      .dive()
    expect(spinner.getElement()).toMatchSnapshot()
  })

  it('should show an error view', async () => {
    const failingProps = {
      ...props,
      findAccount: () => {
        throw new Error('nope')
      }
    }
    const component = shallow(
      <KonnectorModal {...failingProps} />,
      shallowOptions
    )
    await component.instance().componentDidMount()
    component.update()

    const content = component.dive().find('ModalContent')
    expect(content.getElement()).toMatchSnapshot()
  })

  it('should show the configuration view of a single account', async () => {
    const component = shallow(<KonnectorModal {...props} />, shallowOptions)
    await component.instance().componentDidMount()
    component.update()

    const content = component.dive().find('ModalContent')
    expect(content.getElement()).toMatchSnapshot()
  })

  it('should show the list of accounts', async () => {
    const dataBackup = mockKonnector.triggers.data
    mockKonnector.triggers.data = [
      { _id: '784', doctype: 'io.cozy.triggers' },
      { _id: '872', doctype: 'io.cozy.triggers' }
    ]
    const component = shallow(<KonnectorModal {...props} />, shallowOptions)
    await component.instance().componentDidMount()
    component.update()

    const content = component.dive().find('ModalContent')
    expect(content.getElement()).toMatchSnapshot()
    mockKonnector.triggers.data = dataBackup
  })

  it('should render the selected account via a prop', async () => {
    const dataBackup = mockKonnector.triggers.data
    mockKonnector.triggers.data = [
      { _id: '784', doctype: 'io.cozy.triggers' },
      { _id: '872', doctype: 'io.cozy.triggers' }
    ]
    const propsWithSelectedAccount = {
      ...props,
      accountId: '123'
    }
    const component = shallow(
      <KonnectorModal {...propsWithSelectedAccount} />,
      shallowOptions
    )
    await component.instance().componentDidMount()
    component.update()

    const content = component.dive().find('ModalContent')
    expect(content.getElement()).toMatchSnapshot()
    mockKonnector.triggers.data = dataBackup
  })

  describe('adding an account', () => {
    it('should call the parent when controlled by props', () => {
      const createAction = jest.fn()
      const propsWithCreation = {
        ...props,
        createAction
      }
      const component = shallow(
        <KonnectorModal {...propsWithCreation} />,
        shallowOptions
      )
      component.instance().requestAccountCreation()
      expect(createAction).toHaveBeenCalled()
    })

    it('should render the form when controlled by state', async () => {
      const component = shallow(<KonnectorModal {...props} />, shallowOptions)
      await component.instance().componentDidMount()
      component.instance().requestAccountCreation()

      const content = component.dive().find('ModalContent')
      expect(content.getElement()).toMatchSnapshot()
    })
  })

  describe('switching account', () => {
    it('should call the parent when controlled by props', () => {
      const onAccountChange = jest.fn()
      const propsWithSwitch = {
        ...props,
        onAccountChange
      }
      const component = shallow(
        <KonnectorModal {...propsWithSwitch} />,
        shallowOptions
      )
      const account = { _id: '456' }
      const trigger = { _id: 'abc' }
      component.instance().requestAccountChange(account, trigger)
      expect(onAccountChange).toHaveBeenCalledWith(account)
    })

    it('should show the add account view when controlled by state', async () => {
      const component = shallow(<KonnectorModal {...props} />, shallowOptions)
      await component.instance().componentDidMount()
      component.update()

      const account = { _id: '456' }
      const trigger = { _id: 'abc' }
      await component.instance().requestAccountChange(account, trigger)

      const content = component.dive().find('ModalContent')
      expect(content.getElement()).toMatchSnapshot()
    })
  })

  xit('should pass trigger if konnector has triggers', async () => {
    const component = await shallow(
      <KonnectorModal {...props} />,
      shallowOptions
    )
    const manager = component.find(TriggerManager)
    expect(manager.props().trigger).toBeDefined()
  })
})
