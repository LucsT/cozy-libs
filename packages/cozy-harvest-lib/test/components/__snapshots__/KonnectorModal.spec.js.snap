// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KonnectorModal adding an account should render the form when controlled by state 1`] = `
<ModalContent>
  <Wrapper
    konnector={
      Object {
        "name": "Mock",
        "slug": "mock",
        "triggers": Object {
          "data": Array [
            Object {
              "doctype": "io.cozy.triggers",
              "id": 784,
            },
          ],
        },
      }
    }
    onLoginSuccess={[Function]}
    onSuccess={[Function]}
  />
</ModalContent>
`;

exports[`KonnectorModal should show a spinner while loading 1`] = `
<Spinner
  className="u-flex u-flex-justify-center u-pv-3"
  color="var(--primaryColor)"
  loadingType=""
  middle={false}
  noMargin={false}
  size="xxlarge"
/>
`;

exports[`KonnectorModal should show an error view 1`] = `
<ModalContent>
  <Infos
    actionButton={
      <DefaultButton
        theme="danger"
      >
        modal.konnector.error.button
      </DefaultButton>
    }
    icon={null}
    isImportant={true}
    text="modal.konnector.error.description"
    title="modal.konnector.error.title"
  />
</ModalContent>
`;

exports[`KonnectorModal should show the configuration view of a single account 1`] = `
<ModalContent>
  <Wrapper
    account={
      Object {
        "doctype": "io.cozy.accounts",
        "id": 123,
      }
    }
    addAccount={[Function]}
    konnector={
      Object {
        "name": "Mock",
        "slug": "mock",
        "triggers": Object {
          "data": Array [
            Object {
              "doctype": "io.cozy.triggers",
              "id": 784,
            },
          ],
        },
      }
    }
    onAccountDeleted={[MockFunction]}
    refetchTrigger={[Function]}
    trigger={
      Object {
        "doctype": "io.cozy.triggers",
        "id": 784,
      }
    }
  />
</ModalContent>
`;

exports[`KonnectorModal should show the list of accounts 1`] = `
<ModalContent>
  <AccountsList
    accounts={
      Array [
        Object {
          "account": Object {
            "doctype": "io.cozy.accounts",
            "id": 123,
          },
          "trigger": Object {
            "doctype": "io.cozy.triggers",
            "id": 784,
          },
        },
        Object {
          "account": Object {
            "doctype": "io.cozy.accounts",
            "id": 123,
          },
          "trigger": Object {
            "doctype": "io.cozy.triggers",
            "id": 872,
          },
        },
      ]
    }
    addAccount={[Function]}
    konnector={
      Object {
        "name": "Mock",
        "slug": "mock",
        "triggers": Object {
          "data": Array [
            Object {
              "doctype": "io.cozy.triggers",
              "id": 784,
            },
            Object {
              "doctype": "io.cozy.triggers",
              "id": 872,
            },
          ],
        },
      }
    }
    onPick={[Function]}
  />
</ModalContent>
`;
